ethstorage/prover/kzg_poseidon_prover.go

```diff
@@ -4,27 +4,77 @@
 package prover
 
 import (
+	"errors"
 	"fmt"
 	"math/big"
+	"os"
+	"path/filepath"
 
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/log"
 )
 
+type IZKProver interface {
+	// Generate a zk proof for the given encoding keys and samples and return proof and masks (mode 2)
+	GenerateZKProof(encodingKeys []common.Hash, sampleIdxs []uint64) ([]byte, []*big.Int, error)
+	// Generate a zk proof for a given sample (mode 1)
+	GenerateZKProofPerSample(encodingKey common.Hash, sampleIdx uint64) ([]byte, *big.Int, error)
+	// Generate ZK Proof for the given encoding keys and samples and return proof and all publics (mode 2)
+	GenerateZKProofRaw(encodingKeys []common.Hash, sampleIdxs []uint64) ([]byte, []*big.Int, error)
+}
+
+const (
+	SnarkLib  = "snark_lib"
+	WasmName  = "blob_poseidon.wasm"
+	Wasm2Name = "blob_poseidon2.wasm"
+)
+
 type KZGPoseidonProver struct {
-	dir, zkey    string
 	zkProverMode uint64
+	zkProverImpl uint64
+	libDir       string
+	zkey         string
 	lg           log.Logger
 }
 
 // Prover that can be used directly by miner to prove both KZG and Poseidon hash
 // workingDir specifies the working directory of the command relative to the caller.
-// zkeyFileName specifies the zkey file name used by snarkjs to generate snark proof
+// zkeyFileName specifies the zkey file name to generate snark proof
+// zkProverMode specifies the mode of the zk prover, 1 for per sample, 2 for samples
+// zkProverImpl specifies the implementation of the snark prover, 1 for snarkjs, 2 for go-rapidsnark
+// lg specifies the logger to log the info
 // returns a prover that can generate a combined KZG + zk proof
-func NewKZGPoseidonProver(workingDir, zkeyFileName string, mode uint64, lg log.Logger) KZGPoseidonProver {
+func NewKZGPoseidonProverWithLogger(workingDir, zkeyFileName string, zkProverMode, zkProverImpl uint64, lg log.Logger) KZGPoseidonProver {
```

Comment:
What is different between NewKZGPoseidonProver and NewKZGPoseidonProverWithLogger?